### PR TITLE
Feat: Improve existing syntax highlighting

### PR DESCRIPTION
**Pull request by WilsonZiweiWang**
Improve existing syntax highlighting

### CODE REVIEW

File: client/syntaxes/bitbake.tmLanguage.json [hunk 1]
```diff
@@ -1,270 +1,203 @@
...
+            "patterns": [
+                {
+                    "match": "^\\s*(include|require|inherit|export|echo|if|fi|else|return|EXPORT_FUNCTIONS)(?=\\s+|\\s*:)",
+                    "captures": {
+                        "1": {
+                            "name": "keyword.control.bb"
+                        }
+                    }
+                },
+                {
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:51]: Depending on how you want to handle the behaviour with the ":" after the keywords, it might be possible to remove this object and add the "fakeroot" keyword to the list at line 44

File: client/syntaxes/bitbake.tmLanguage.json [hunk 2]
```diff
@@ -1,270 +1,203 @@
...
+                    "end": "(\")",
+                    "patterns": [
+                        {
+                            "include": "#variable-reference"
+                        }
+                    ]
+                },
+                {
+                    "name": "string.quoted.single.bb",
+                    "begin": "'",
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:91]: Replace with "(')" for more code consistency.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 3]
```diff
@@ -1,270 +1,203 @@
...
+                    "end": "'",
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:92]: Replace with "(')" for more code consistency.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 4]
```diff
@@ -1,270 +1,203 @@
...
+                }
+            },
+            "end": "(?=\\))",
+            "patterns": [
+                {
+                    "include": "#function-params"
+                }  
+            ]
+        },
+        "function-content": {
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:141]: The "function-content" object has no effect. The included patterns are already included in the patterns array at the top level (line 11-39). Therefore they will get highlighted even without this block of code.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 5]
```diff
@@ -1,270 +1,203 @@
...
+            "begin": "(?<=\\(.*\\)\\{)",
+            "end": "(?=\\})",
+            "patterns": [
+                {
+                    "include": "#variable-assignment"
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:146]: There is no variable-assignment object defined anymore. Therefore this line has no effect and can be removed.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 6]
```diff
@@ -1,270 +1,203 @@
...
+                {
+                    "include": "#string"
+                },
+                {
+                    "include": "#variable-reference"
+                }
+            ]
+        },
+        "variable-reference": {
+            "patterns": [
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:157]: This patterns array can be removed since it contains only one object.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 7]
```diff
@@ -1,270 +1,203 @@
...
+                {
+                    "match": "(\\$\\{)([a-zA-Z0-9_]*)(\\})",
+                    "name": "variable.language.bb",
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:160]: The name on this line and the one on line 163 are redundant. One of them can be removed. If this line gets deleted there must be an additional 3. object in the captures object to highlight the closing bracket.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 8]
```diff
@@ -1,270 +1,203 @@
...
+                        },
+                        "2":{
+                            "name": "variable.other.bb"
+                        }
+                    }
+                }
+            ]
+        },
+        "variable-name": {
+            "match": "\\s*([a-zA-Z0-9_-][\\w-]*)(:append|:prepend:remove)?(?!\\s*\\()",
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:173]: Missing OR operator between ":prepend" and ":remove".

File: client/syntaxes/bitbake.tmLanguage.json [hunk 9]
```diff
@@ -1,270 +1,203 @@
...
+        },
+        "function-params":{
+            "begin": "(?<=\\()",
+            "end": "(?=\\))",
+            "patterns": [
+                {
+                    "include": "#variable-name"
+                },
+                {
+                    "include": "#variable-assignment"
```
Comment (gwerderm) [inline client/syntaxes/bitbake.tmLanguage.json:191]: There is no variable-assignment object defined anymore. Therefore this line has no effect and can be removed.

File: client/syntaxes/bitbake.tmLanguage.json [hunk 10]
```diff
@@ -1,270 +1,203 @@
...
         {
-          "include": "#comment"
+            "include": "#function-content"
+        }
+    ],
+    "repository": {
+        "keywords":{
+            "patterns": [
+                {
+                    "match": "^\\s*(include|require|inherit|export|echo|if|fi|else|return|EXPORT_FUNCTIONS)(?=\\s+|\\s*:)",
```
[the remark this answers is not in the document]
Comment (WilsonZiweiWang) [reply client/syntaxes/bitbake.tmLanguage.json]: From what I know so far, the `else` is the only keyword that can be followed by a `:`. I can create another rule for it.

[the remark this answers is not in the document]
Comment (WilsonZiweiWang) [reply client/syntaxes/bitbake.tmLanguage.json]: Also, I am figuring out if I should always highlight the correct syntax or highlight them when they are present. From my experience, both behaviours could exist in the same extension. I wonder what would be the better option for BitBake users.

**Comment by WilsonZiweiWang** [comment]
> I just stumbled across this pull request and found some things i would do slightly different. Also (I didn't comment this in the code) i was asking myself, what might has been the reason for you to remove the includes of "source.python" and "source.shell". I think using them could make it easier to write the code, while being sure that the code inside functions gets highlighted properly. But maybe I am wrong with that?

Thanks for the feedback, I pushed some new changes before I saw your comments. The new one should have fixed most of the issues you mentioned.

As for `source.python` and `source.shell`, they caused me some trouble at the beginning because some of the rules have conflicts. So I removed them. But you are right that they will save me some effort. To take advantage of them, I figure it would be better to include them in scopes that only Python-like or Shell-like code is present instead of in the global.

**Review by idillon-sfl — APPROVED** [review]
Looks quite good. Just a bunch of corner cases I found:

```
# python should be colored as a variable
python = test

# inherit should be colored as a function name
python inherit () {}

# 'then' should be same color as 'if'
if [ somecondition ] ; then

# no idea what should happen, but this is weird
TEST = python inherit () {}

# This one too is weird
inherit from export inherit from test inherit test from inherit
```

That being said, don't hesitate to just merge and look at it later.

I would also suggest to use a json prettifier to set the indentation to 2 spaces so we can track the changes better.
